pkg/memo/memo_test.go

```diff
@@ -0,0 +1,264 @@
+	fString := "this_is_a_string_field"
+
+	tests := []struct {
+		name         string
+		memo         *memo.InboundMemo
+		expectedHead []byte
+		expectedData []byte
+		errMsg       string
+	}{
+		{
+			name: "encode memo with ABI encoding",
+			memo: &memo.InboundMemo{
+				Header: memo.Header{
+					Version:        0,
+					EncodingFormat: memo.EncodingFmtABI,
+					OpCode:         memo.OpCodeDepositAndCall,
+				},
+				FieldsV0: memo.FieldsV0{
+					Receiver: fAddress,
+					Payload:  fBytes,
+					RevertOptions: crosschaintypes.RevertOptions{
+						RevertAddress: fString,
+						CallOnRevert:  true,
+						AbortAddress:  fAddress.String(), // it's a ZEVM address
+						RevertMessage: fBytes,
+					},
+				},
+			},
+			expectedHead: sample.MemoHead(
+				0,
+				memo.EncodingFmtABI,
+				memo.OpCodeDepositAndCall,
+				0,
+				flagsAllFieldsSet, // all fields are set
+			),
+			expectedData: sample.ABIPack(t,
+				memo.ArgReceiver(fAddress),
+				memo.ArgPayload(fBytes),
+				memo.ArgRevertAddress(fString),
+				memo.ArgAbortAddress(fAddress),
+				memo.ArgRevertMessage(fBytes)),
+		},
+		{
+			name: "encode memo with compact encoding",
+			memo: &memo.InboundMemo{
+				Header: memo.Header{
+					Version:        0,
+					EncodingFormat: memo.EncodingFmtCompactShort,
+					OpCode:         memo.OpCodeDepositAndCall,
+				},
+				FieldsV0: memo.FieldsV0{
+					Receiver: fAddress,
+					Payload:  fBytes,
+					RevertOptions: crosschaintypes.RevertOptions{
+						RevertAddress: fString,
+						CallOnRevert:  true,
+						AbortAddress:  fAddress.String(), // it's a ZEVM address
+						RevertMessage: fBytes,
+					},
+				},
+			},
+			expectedHead: sample.MemoHead(
+				0,
+				memo.EncodingFmtCompactShort,
+				memo.OpCodeDepositAndCall,
+				0,
+				flagsAllFieldsSet, // all fields are set
+			),
+			expectedData: sample.CompactPack(
+				memo.EncodingFmtCompactShort,
+				memo.ArgReceiver(fAddress),
+				memo.ArgPayload(fBytes),
+				memo.ArgRevertAddress(fString),
+				memo.ArgAbortAddress(fAddress),
+				memo.ArgRevertMessage(fBytes)),
+		},
+		{
+			name: "failed to encode memo header",
+			memo: &memo.InboundMemo{
+				Header: memo.Header{
+					OpCode: memo.OpCodeInvalid, // invalid operation code
+				},
+			},
+			errMsg: "failed to encode memo header",
+		},
+		{
+			name: "failed to encode if version is invalid",
+			memo: &memo.InboundMemo{
+				Header: memo.Header{
+					Version: 1,
+				},
+			},
+			errMsg: "invalid memo version",
+		},
+		{
+			name: "failed to pack memo fields",
+			memo: &memo.InboundMemo{
+				Header: memo.Header{
+					Version:        0,
+					EncodingFormat: memo.EncodingFmtABI,
+					OpCode:         memo.OpCodeDeposit,
+				},
+				FieldsV0: memo.FieldsV0{
+					Receiver: fAddress,
+					Payload:  fBytes, // payload is not allowed for deposit
+				},
+			},
+			errMsg: "failed to pack memo fields version: 0",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
Let's also add a decode step as well to ensure that `before{} == after{}`

Reply:
good idea. added https://github.com/zeta-chain/node/pull/2987/commits/d022be1053ebba9c11d5d3d5c51755b3705b4c03